src/default-certhash-codec.ts

```diff
@@ -0,0 +1,45 @@
+import { ParseError } from './codec.js'
+import type { Codec } from './protocols-table'
+import { concat as uint8ArrayConcat } from 'uint8arrays/concat'
+import { toString as uint8ArrayToString } from 'uint8arrays/to-string'
+import varint  from 'varint'
+import { base16 } from 'multiformats/bases/base16'
+import { base58btc } from 'multiformats/bases/base58'
+import { base64url } from 'multiformats/bases/base64'
+import type { BaseDecoder } from 'multiformats/bases/interface'
+
+const encodings: {[prefix: string]: BaseDecoder} = {
+    'f': base16,
+    'u': base64url,
+    'z': base58btc,
+};
```

Comment:
I am not familiar with the JS multiformats stack. I am surprised that this mapping needs to be defined. I would expect this to be a general multihash and not a certhash specific mapping is.

Reply:
You can compose a decoder from several other decoders, though the [types are a little wonky](https://github.com/multiformats/js-multiformats/issues/141).

> I would expect this to be a general multihash

Do you mean not limited to the `f`, `u` or `z` prefixes?  `multiformat`s follows a pattern where you only include the codecs you need instead of the kitchen-sink approach.

You can, of course, compose your codec out of all available codecs.

Reply:
> > I would expect this to be a general multihash
> 
> Do you mean not limited to the `f`, `u` or `z` prefixes? `multiformat`s follows a pattern where you only include the codecs you need instead of the kitchen-sink approach.

I was not referring to supporting all multihash variants, but instead to the fact that the mapping between identifier (e.g. `f`) and its multihash variant (here `base16`) needs to be written by the user. That seems error prone to me. I might just be missing something.

> And to be clear, we're saying we want to support _all_ of them, to take the draft spec as currently written literally, regardless of who is using this library? Except, perhaps, encodings with '/' in their alphabet?

That is how we implement it in Rust https://github.com/multiformats/rust-multiaddr/pull/59 and Golang https://github.com/multiformats/go-multiaddr/pull/176/. That said, both encode with `Base64Url`. Thus I think it is fine to support only a minimal set of bases for now.

We might require users to use sha256. In the same vein, we could as well require users to use `Base64Url` for now. Thoughts?

Reply:
> the mapping between identifier (e.g. f) and its multihash variant (here base16) needs to be written by the user.
It doesn't and if you see the new version of the code I dropped that.

> Base64Url. Thus I think it is fine to support only a minimal set of bases for now.
I'll do whichever you like. Right now it's back to supporting "all" of them.

I put all in quotes because of the '/' issue. I didn't see an easy way to detect the alphabets from the dictionary provided by the `multiformats` library so I didn't filter them out. I would just guess that a base64-encoded certhash that happens to have a / in it, if placed in a Multiaddr, would cause problems at runtime.

Reply:
For the record https://github.com/libp2p/specs/pull/412/commits/7009f943d5b9ee67cb7a19bbf8ca60549c1ad683 specifies the requirement of supporting `sha256` and `base64url`. I don't have an opinion on whether js-multiaddr should support any other algorithms / encodings beyond that.

Reply:
Well-spotted. If the other side is only required to accept those two, then base58btc was a very poor choice of encoding for convertToString(). I'll fix that.

I think we can afford to leave the convertToBytes as supporting a bunch of 'em, since I'm not adding anything new to package.json to get it. But I'll certainly remove those others if advised to do so.